substrate-node/pallets/pallet-dao/src/dao.rs

```diff
@@ -0,0 +1,366 @@
+
+        let p = proposal::DaoProposal {
+            index,
+            description,
+            link,
+        };
+        <Proposals<T>>::insert(proposal_hash, p);
+
+        let votes = {
+            proposal::DaoVotes {
+                index,
+                threshold,
+                ayes: vec![],
+                nays: vec![],
+                end,
+                vetos: vec![],
+            }
+        };
+        <Voting<T>>::insert(proposal_hash, votes);
+
+        let mut active_proposal_hashes = <ProposalList<T>>::get();
+        active_proposal_hashes.push(proposal_hash);
+        <ProposalList<T>>::set(active_proposal_hashes);
+
+        Self::deposit_event(Event::Proposed {
+            account: who,
+            proposal_index: index,
+            proposal_hash,
+            threshold,
+        });
+
+        Ok(().into())
+    }
+
+    pub fn _vote(
+        who: T::AccountId,
+        farm_id: u32,
+        proposal_hash: T::Hash,
+        approve: bool,
+    ) -> DispatchResultWithPostInfo {
+        ensure!(
+            T::Tfgrid::is_farm_owner(farm_id, who.clone()),
+            Error::<T>::NotAuthorizedToVote
+        );
+
+        let stored_proposal =
+            <Proposals<T>>::get(proposal_hash).ok_or(Error::<T>::ProposalMissing)?;
+
+        let mut voting = Self::voting(proposal_hash).ok_or(Error::<T>::ProposalMissing)?;
+        ensure!(
+            voting.index == stored_proposal.index,
+            Error::<T>::WrongIndex
+        );
+
+        // Don't allow votes after the time limit reached
+        ensure!(
+            frame_system::Pallet::<T>::block_number() <= voting.end,
+            Error::<T>::TimeLimitReached
+        );
+
+        let position_yes = voting.ayes.iter().position(|a| a.farm_id == farm_id);
+        let position_no = voting.nays.iter().position(|a| a.farm_id == farm_id);
+
+        // Detects first vote of the member in the motion
+        let is_account_voting_first_time = position_yes.is_none() && position_no.is_none();
+
+        if approve {
+            if position_yes.is_none() {
+                voting.ayes.push(proposal::VoteWeight {
+                    farm_id: farm_id,
+                    weight: Self::get_vote_weight(farm_id)?,
+                });
+            } else {
+                return Err(Error::<T>::DuplicateVote.into());
+            }
+            if let Some(pos) = position_no {
+                voting.nays.swap_remove(pos);
+            }
+        } else {
+            if position_no.is_none() {
+                voting.nays.push(proposal::VoteWeight {
+                    farm_id: farm_id,
+                    weight: Self::get_vote_weight(farm_id)?,
+                });
+            } else {
+                return Err(Error::<T>::DuplicateVote.into());
+            }
+            if let Some(pos) = position_yes {
+                voting.ayes.swap_remove(pos);
+            }
+        }
+
+        let yes_votes = voting.ayes.len() as u32;
```

Comment:
Nitpick: no need to create new variables here. We can just move the right side of the operation in the deposit_event call.